---
title: "R Notebook"
output: html_notebook
---
SOMEWHAT FULL SCRIPT 


**Workflow**:

  1. *Subsetting the data.R*: load, subset & sample dataset 
      * Output: countries (dataframe)
      
  2. *Fasta_function.R*: initialize FASTA manipulating functions
  
  3. *spike_reference_test.R*: Load in reference spike sequence and get start/end motifs.
  
  4. *Full Data.R*: Get FASTA files, process & isolate target sequences, align sequences & flag mutations
      * Output: df_isolate (dataframe)
      
  5. *Mutation_function.R*: Initialize mutation isolation functions, load and format reference mutation data
      * Output: mutations_ref (dataframe)
  
  .
  .
  .

More to follow...

### Loading libraries
```{r}
library(ape)
library(adegenet)
library(dplyr)
library(stringr)
library(readr)
library(rentrez)
```


# Importing and fixing covid sequences dataset
```{r}
# Loading Sequence data & filtering by completeness, trimming un-used cols
dat <- read.csv('./Dataset/sequences.csv') %>%
  filter(Nuc_Completeness == "complete") %>%
  transmute(Accession, Length, Geo_Location)

# Setting number of samples per country:
N <- 50

# Filter dataset by country, and sampling N random samples from each
dat_USA <- dat %>%
  filter(str_detect(Geo_Location, 'USA')) %>%
  mutate(Geo_Location = "USA") %>%
  slice_sample(n = N)

dat_CHINA <- dat %>%
  filter(str_detect(Geo_Location, 'China')) %>%
  mutate(Geo_Location = "CHN") %>%
  slice_sample(n = N)

dat_AUSTRALIA <- dat %>%
  filter(str_detect(Geo_Location, 'Australia')) %>%
  mutate(Geo_Location = "AUS") %>%
  slice_sample(n = N)

dat_INDIA <- dat %>%
  filter(str_detect(Geo_Location, 'India')) %>%
  mutate(Geo_Location = "IND") %>%
  slice_sample(n = N)

dat_JAPAN <- dat %>%
  filter(str_detect(Geo_Location, 'Japan')) %>%
  mutate(Geo_Location = "JPN") %>%
  slice_sample(n = N)

# Dataset with the selected countries
countries <- rbind(dat_JAPAN, dat_INDIA, dat_AUSTRALIA, dat_CHINA, dat_USA)

# Removing the temp data files
rm(dat, N, dat_AUSTRALIA, dat_CHINA, dat_USA, dat_INDIA, dat_JAPAN)
```

# Importing data frame with the known mutations
```{r}
mutations <- read.csv('./Dataset/mutations.csv')
SpikeStart<-21563
mutations$RelativeLocation<-mutations$Location- SpikeStart
```

# Importing FASTA file of the reference spike protein nucleotide
```{r}
# Loading in Fasta_functions.R file
source("./Fasta_functions.R")

####### PREPARING SPIKE TEMPLATE #######

# Setting correct filename & extension
fileName <- "./Dataset/spike_reference.txt"

# Load in the reference sequence
spike_fasta <-  readChar(fileName, file.info(fileName)$size)

# Clean FASTA file & remove sequence
spike_seq <- separate_fasta(spike_fasta)$Sequence

# Getting start & end motifs to capture our sequence of 12 bp/side
start_motif <- substr(spike_seq, 1, 12)
end_motif <- substr(spike_seq, nchar(spike_seq) - 11, nchar(spike_seq))
```

# Loading Spike reference test
```{r}
# Setting correct filename & extension
fileName <- "./Dataset/spike_reference.txt"

# Load in the reference sequence
spike_fasta <-  readChar(fileName, file.info(fileName)$size)

# Clean FASTA file & convert to format compatible with dataset
spike_ref <- separate_fasta(spike_fasta) %>%
  mutate(Target = Sequence,
         Sequence = "NULL",
         seq_len = nchar(Target),
         Geo_Location = "REF")

# Getting start & end motifs to capture our sequence of 12 bp/side
start_motif <- substr(spike_ref$Target, 1, 12)
end_motif <- substr(spike_ref$Target,
                    nchar(spike_ref$Target) - 11,
                    nchar(spike_ref$Target))

# Removing excess variables
rm(fileName, spike_fasta)
```












